File: nodejs/test/commands.test.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, expect, it, onTestFinished, vi, beforeEach } from "vitest";
import { approveAll, CopilotClient } from "../src/index.js";
import { CommandRegistry, BUILTIN_COMMANDS } from "../src/commands.js";

describe("SlashCommand Types", () => {
    it("BUILTIN_COMMANDS contains expected commands", () => {
        const commandNames = BUILTIN_COMMANDS.map((c) => c.name);

        expect(commandNames).toContain("/help");
        expect(commandNames).toContain("/model");
        expect(commandNames).toContain("/compact");
        expect(commandNames).toContain("/clear");
        expect(commandNames).toContain("/agent");

        for (const cmd of BUILTIN_COMMANDS) {
            expect(cmd.name).toMatch(/^\//);
            expect(cmd.description).toBeTruthy();
            expect(typeof cmd.description).toBe("string");
        }
    });

    it("each command has required SlashCommand shape", () => {
        for (const cmd of BUILTIN_COMMANDS) {
            expect(cmd).toHaveProperty("name");
            expect(cmd).toHaveProperty("description");
            // parameters is optional
            if (cmd.parameters) {
                expect(Array.isArray(cmd.parameters)).toBe(true);
                for (const param of cmd.parameters) {
                    expect(param).toHaveProperty("name");
                    expect(param).toHaveProperty("description");
                    expect(param).toHaveProperty("required");
                }
            }
        }
    });
});

describe("CommandRegistry", () => {
    let registry: CommandRegistry;

    beforeEach(() => {
        registry = new CommandRegistry();
    });

    it("list() returns all builtin commands", () => {
        const commands = registry.list();
        expect(commands.length).toBeGreaterThan(0);
        expect(commands.length).toBe(BUILTIN_COMMANDS.length);
    });

    it("get() returns a command by name", () => {
        const cmd = registry.get("/help");
        expect(cmd).toBeDefined();
        expect(cmd!.name).toBe("/help");
    });

    it("get() returns undefined for unknown command", () => {
        const cmd = registry.get("/nonexistent");
        expect(cmd).toBeUndefined();
    });

    it("has() returns true for known commands", () => {
        expect(registry.has("/help")).toBe(true);
        expect(registry.has("/model")).toBe(true);
    });

    it("has() returns false for unknown commands", () => {
        expect(registry.has("/nonexistent")).toBe(false);
    });

    it("hasRpcMapping() identifies commands with RPC equivalents", () => {
        expect(registry.hasRpcMapping("/model")).toBe(true);
        expect(registry.hasRpcMapping("/compact")).toBe(true);
        expect(registry.hasRpcMapping("/agent")).toBe(true);
    });

    it("hasRpcMapping() returns false for passthrough commands", () => {
        expect(registry.hasRpcMapping("/help")).toBe(false);
        expect(registry.hasRpcMapping("/nonexistent")).toBe(false);
    });

    it("parseCommandLine() parses command name and args", () => {
        expect(registry.parseCommandLine("/help")).toEqual({
            name: "/help",
            args: [],
        });
        expect(registry.parseCommandLine("/model gpt-4")).toEqual({
            name: "/model",
            args: ["gpt-4"],
        });
        expect(registry.parseCommandLine("/agent select my-agent")).toEqual({
            name: "/agent",
            args: ["select", "my-agent"],
        });
    });

    it("parseCommandLine() handles leading/trailing whitespace", () => {
        expect(registry.parseCommandLine("  /help  ")).toEqual({
            name: "/help",
            args: [],
        });
    });

    it("parseCommandLine() returns null for non-command text", () => {
        expect(registry.parseCommandLine("hello world")).toBeNull();
        expect(registry.parseCommandLine("")).toBeNull();
    });
});

describe("CopilotSession slash commands", () => {
    it("supportedCommands() returns list of slash commands", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const commands = session.supportedCommands();
        expect(commands.length).toBeGreaterThan(0);
        for (const cmd of commands) {
            expect(cmd.name).toMatch(/^\//);
            expect(cmd.description).toBeTruthy();
        }
    });

    it("sendCommand() calls RPC for /model command", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const spy = vi.spyOn(session.rpc.model, "switchTo");
        spy.mockResolvedValue({} as any);

        const result = await session.sendCommand("/model", ["gpt-4"]);
        expect(result.handled).toBe(true);
        expect(result.method).toBe("rpc");
        expect(spy).toHaveBeenCalledWith(expect.objectContaining({ modelId: "gpt-4" }));
    });

    it("sendCommand() calls RPC for /compact command", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const spy = vi.spyOn(session.rpc.compaction, "compact");
        spy.mockResolvedValue({} as any);

        const result = await session.sendCommand("/compact");
        expect(result.handled).toBe(true);
        expect(result.method).toBe("rpc");
        expect(spy).toHaveBeenCalled();
    });

    it("sendCommand() calls RPC for /agent select", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const spy = vi.spyOn(session.rpc.agent, "select");
        spy.mockResolvedValue({} as any);

        const result = await session.sendCommand("/agent", ["select", "my-agent"]);
        expect(result.handled).toBe(true);
        expect(result.method).toBe("rpc");
        expect(spy).toHaveBeenCalledWith(expect.objectContaining({ name: "my-agent" }));
    });

    it("sendCommand() falls back to send() for unknown commands", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const spy = vi.spyOn(session, "send");
        spy.mockResolvedValue("msg-123");

        const result = await session.sendCommand("/some-unknown-command", ["arg1"]);
        expect(result.handled).toBe(true);
        expect(result.method).toBe("passthrough");
        expect(spy).toHaveBeenCalledWith(
            expect.objectContaining({ prompt: "/some-unknown-command arg1" })
        );
    });

    it("sendCommand() returns help text for /help", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const result = await session.sendCommand("/help");
        expect(result.handled).toBe(true);
        expect(result.method).toBe("local");
        expect(result.output).toBeTruthy();
        expect(result.output).toContain("/help");
        expect(result.output).toContain("/model");
    });

    it("sendCommand() handles /clear via local handler", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const result = await session.sendCommand("/clear");
        expect(result.handled).toBe(true);
        expect(result.method).toBe("local");
    });

    it("sendCommand() parses full command string", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const spy = vi.spyOn(session.rpc.model, "switchTo");
        spy.mockResolvedValue({} as any);

        // Can also pass the full command string
        const result = await session.sendCommand("/model gpt-4");
        expect(result.handled).toBe(true);
        expect(spy).toHaveBeenCalledWith(expect.objectContaining({ modelId: "gpt-4" }));
    });

    it("sendCommand() throws for empty command", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        await expect(session.sendCommand("")).rejects.toThrow(/invalid.*command/i);
    });

    it("sendCommand() throws for non-slash-prefixed input", async () => {
        const client = new CopilotClient();
        await client.start();
        onTestFinished(() => client.forceStop());

        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        await expect(session.sendCommand("hello")).rejects.toThrow(/must start with/i);
    });
});

File: go/commands_test.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package copilot

import (
	"testing"
)

func TestBuiltinCommands(t *testing.T) {
	t.Run("contains expected commands", func(t *testing.T) {
		names := make(map[string]bool)
		for _, cmd := range builtinCommands {
			names[cmd.Name] = true
		}

		for _, expected := range []string{"/help", "/model", "/compact", "/clear", "/agent"} {
			if !names[expected] {
				t.Errorf("expected builtin command %q not found", expected)
			}
		}
	})

	t.Run("all commands start with slash", func(t *testing.T) {
		for _, cmd := range builtinCommands {
			if cmd.Name[0] != '/' {
				t.Errorf("command %q does not start with /", cmd.Name)
			}
			if cmd.Description == "" {
				t.Errorf("command %q has empty description", cmd.Name)
			}
		}
	})
}

func TestCommandRegistry(t *testing.T) {
	registry := NewCommandRegistry()

	t.Run("List returns all builtin commands", func(t *testing.T) {
		commands := registry.List()
		if len(commands) != len(builtinCommands) {
			t.Errorf("expected %d commands, got %d", len(builtinCommands), len(commands))
		}
	})

	t.Run("Get returns command by name", func(t *testing.T) {
		cmd := registry.Get("/help")
		if cmd == nil {
			t.Fatal("expected /help command, got nil")
		}
		if cmd.Name != "/help" {
			t.Errorf("expected name /help, got %s", cmd.Name)
		}
	})

	t.Run("Get returns nil for unknown command", func(t *testing.T) {
		cmd := registry.Get("/nonexistent")
		if cmd != nil {
			t.Errorf("expected nil for unknown command, got %v", cmd)
		}
	})

	t.Run("Has returns true for known commands", func(t *testing.T) {
		if !registry.Has("/help") {
			t.Error("expected Has(/help) to be true")
		}
		if !registry.Has("/model") {
			t.Error("expected Has(/model) to be true")
		}
	})

	t.Run("Has returns false for unknown commands", func(t *testing.T) {
		if registry.Has("/nonexistent") {
			t.Error("expected Has(/nonexistent) to be false")
		}
	})

	t.Run("HasRpcMapping for RPC commands", func(t *testing.T) {
		if !registry.HasRpcMapping("/model") {
			t.Error("expected HasRpcMapping(/model) to be true")
		}
		if !registry.HasRpcMapping("/compact") {
			t.Error("expected HasRpcMapping(/compact) to be true")
		}
		if !registry.HasRpcMapping("/agent") {
			t.Error("expected HasRpcMapping(/agent) to be true")
		}
	})

	t.Run("HasRpcMapping false for passthrough commands", func(t *testing.T) {
		if registry.HasRpcMapping("/help") {
			t.Error("expected HasRpcMapping(/help) to be false")
		}
		if registry.HasRpcMapping("/nonexistent") {
			t.Error("expected HasRpcMapping(/nonexistent) to be false")
		}
	})

	t.Run("ParseCommandLine basic", func(t *testing.T) {
		result := registry.ParseCommandLine("/help")
		if result == nil {
			t.Fatal("expected parsed result, got nil")
		}
		if result.Name != "/help" {
			t.Errorf("expected name /help, got %s", result.Name)
		}
		if len(result.Args) != 0 {
			t.Errorf("expected 0 args, got %d", len(result.Args))
		}
	})

	t.Run("ParseCommandLine with args", func(t *testing.T) {
		result := registry.ParseCommandLine("/model gpt-4")
		if result == nil {
			t.Fatal("expected parsed result, got nil")
		}
		if result.Name != "/model" {
			t.Errorf("expected name /model, got %s", result.Name)
		}
		if len(result.Args) != 1 || result.Args[0] != "gpt-4" {
			t.Errorf("expected args [gpt-4], got %v", result.Args)
		}
	})

	t.Run("ParseCommandLine multiple args", func(t *testing.T) {
		result := registry.ParseCommandLine("/agent select my-agent")
		if result == nil {
			t.Fatal("expected parsed result, got nil")
		}
		if result.Name != "/agent" {
			t.Errorf("expected name /agent, got %s", result.Name)
		}
		if len(result.Args) != 2 || result.Args[0] != "select" || result.Args[1] != "my-agent" {
			t.Errorf("expected args [select my-agent], got %v", result.Args)
		}
	})

	t.Run("ParseCommandLine with whitespace", func(t *testing.T) {
		result := registry.ParseCommandLine("  /help  ")
		if result == nil {
			t.Fatal("expected parsed result, got nil")
		}
		if result.Name != "/help" {
			t.Errorf("expected name /help, got %s", result.Name)
		}
	})

	t.Run("ParseCommandLine non-command returns nil", func(t *testing.T) {
		result := registry.ParseCommandLine("hello world")
		if result != nil {
			t.Errorf("expected nil for non-command, got %v", result)
		}

		result = registry.ParseCommandLine("")
		if result != nil {
			t.Errorf("expected nil for empty string, got %v", result)
		}
	})
}

func TestSession_SupportedCommands(t *testing.T) {
	session, cleanup := newTestSession()
	defer cleanup()

	commands := session.SupportedCommands()
	if len(commands) == 0 {
		t.Fatal("expected non-empty commands list")
	}

	for _, cmd := range commands {
		if cmd.Name[0] != '/' {
			t.Errorf("command %q does not start with /", cmd.Name)
		}
		if cmd.Description == "" {
			t.Errorf("command %q has empty description", cmd.Name)
		}
	}
}

func TestSession_SendCommand_Validation(t *testing.T) {
	session, cleanup := newTestSession()
	defer cleanup()

	t.Run("empty command returns error", func(t *testing.T) {
		_, err := session.SendCommand(t.Context(), "")
		if err == nil {
			t.Error("expected error for empty command")
		}
	})

	t.Run("non-slash command returns error", func(t *testing.T) {
		_, err := session.SendCommand(t.Context(), "hello")
		if err == nil {
			t.Error("expected error for non-slash command")
		}
	})

	t.Run("help command returns local result", func(t *testing.T) {
		result, err := session.SendCommand(t.Context(), "/help")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !result.Handled {
			t.Error("expected handled to be true")
		}
		if result.Method != "local" {
			t.Errorf("expected method local, got %s", result.Method)
		}
		if result.Output == "" {
			t.Error("expected non-empty output for /help")
		}
	})

	t.Run("clear command returns local result", func(t *testing.T) {
		result, err := session.SendCommand(t.Context(), "/clear")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !result.Handled {
			t.Error("expected handled to be true")
		}
		if result.Method != "local" {
			t.Errorf("expected method local, got %s", result.Method)
		}
	})
}
